dex/wait/queue.go

```diff
@@ -50,47 +58,48 @@ func NewTickerQueue(recheckInterval time.Duration) *TickerQueue {
 // Wait attempts to run the (*Waiter).TryFunc until either 1) the function
 // returns the value DontTryAgain, or 2) the function's Expiration time has
 // passed. In the case of 2, the (*Waiter).ExpireFunc will be run.
-func (w *TickerQueue) Wait(settings *Waiter) {
-	if time.Now().After(settings.Expiration) {
+func (q *TickerQueue) Wait(w *Waiter) {
+	if time.Now().After(w.Expiration) {
 		log.Error("wait.TickerQueue: Waiter given expiration before present")
 		return
 	}
 	// Check to see if it passes right away.
-	if settings.TryFunc() {
+	if w.TryFunc() == DontTryAgain {
 		return
 	}
-	w.waiterMtx.Lock()
-	w.waiters = append(w.waiters, settings)
-	w.waiterMtx.Unlock()
+	q.waiterMtx.Lock()
+	q.waiters = append(q.waiters, w)
+	q.waiterMtx.Unlock()
 }
 
 // Run runs the primary wait loop until the context is canceled.
-func (w *TickerQueue) Run(ctx context.Context) {
+func (q *TickerQueue) Run(ctx context.Context) {
 	// The latencyTicker triggers a check of all waitFunc functions.
-	latencyTicker := time.NewTicker(w.recheckInterval)
+	latencyTicker := time.NewTicker(q.recheckInterval)
 	defer latencyTicker.Stop()
 
 	runWaiters := func() {
-		w.waiterMtx.Lock()
-		defer w.waiterMtx.Unlock()
+		q.waiterMtx.Lock()
+		defer q.waiterMtx.Unlock()
 		agains := make([]*Waiter, 0)
 		// Grab new waiters
 		tNow := time.Now()
-		for _, settings := range w.waiters {
+		for _, w := range q.waiters {
 			if ctx.Err() != nil {
 				return
 			}
-			if !settings.TryFunc() {
-				// If this waiter has expired, issue the timeout error to the client
-				// and do not append to the agains slice.
-				if settings.Expiration.Before(tNow) {
-					settings.ExpireFunc()
-					continue
-				}
-				agains = append(agains, settings)
-			} // End if !mFunc.f(). Nothing to do if mFunc returned dontTryAgain=true.
+			if w.TryFunc() == DontTryAgain {
```

Comment:
Excellent to make this a type. Treating it as a bool was really confusing.  I wouldn't even object to the underlying type for `TryDirective` being an int to prevent ever using it as a bool, but this is still good.